### DISCUSSION TITLE
Headers Error in 7-zip

### DISCUSSION THREAD

**Discussion by hrishikesh-k** [101arrowz/fflate#92] [Q&A] [answered]
UPDATE: Looks like this problem exists only when unzipping with 7-zip - however the archive unzips fine and the files are also accessible.

Hello! Thank you for this wonderful package! I've just run into a problem, hoping if I can get some help.

I was trying to implement a solution to fetch files from API and add them to a zip (simply store - no compression). I am receiving a list of all files from an API - I'm then downloading the first file, adding it to zip, then downloading the next file, adding it to zip and so on... I was able to follow this: https://github.com/101arrowz/fflate/wiki/Guide:-Modern-(Buildless)

I'm doing this:

```js
// callApi is my custom function using browser's fetch API that returns `response.blob()`
// files is the array of file list from API. It looks like [{path:'foo'},{path:'bar'}]

import {Zip, ZipPassThrough} from 'fflate'

downloadAll(index, zip, zipReadableStream) {

  const fflToRS = fflateStream => new ReadableStream({
    start(controller) {
      fflateStream.ondata = (error, data, final) => {
        if (error) {
          controller.error(error)
        } else {
          controller.enqueue(data)
          if (final) {
            controller.close()
          }
        }
      }
    },
    cancel() {
      fflateStream.terminate()
    }
  })

  const addFileToZip = async (fileData, fileObject, zip) => {
    const zippedFileStream = new ZipPassThrough(fileObject.path)
    zip.add(zippedFileStream)
    const fileReader = fileData.stream().getReader()
    while (true) {
      const {done, value} = await fileReader.read()
      if (done) {
        zippedFileStream.push(new Uint8Array(0), true)
        break
      }
      zippedFileStream.push(value)
    }
  }

  zip = zip ? zip : new Zip()
  zipReadableStream = zipReadableStream ? zipReadableStream : fflToRS(zip)

  callApi('url', true).then(file => {
    addFileToZip(file, files[index], zip).then(() => {
      if (index++ < files.length - 1) {
        downloadAll(index++, zip, zipReadableStream)
      } else {
        zip.end()
        const zipResponse = new Response(zipReadableStream)
        // rest
      }
    })
  })

}
```
I'm calling the function for the first time like `downloadAll(0)`. I get the zip as needed, but when I try to extract it with 7-zip it gives me this error:

![image](https://user-images.githubusercontent.com/11403551/137688825-811fd639-e7c1-486c-9627-dbf86cc4d99e.png)

Just wondering if there's any way to get rid of that error/warning.

**Comment by 101arrowz** [comment, marked as answer]
Does 7-zip provide any more details than "Headers Error"? Also, could you send a ZIP file that reproduces this?

The fact it unzips properly despite the "error" indicates to me that there might be some kind of incompatibility. Whether or not it's something `fflate` can resolve I'd like to investigate. It seems similar to [this comment](https://github.com/barracudanetworks/ArchiveStream-php/issues/44#issuecomment-464391412).

**Reply by hrishikesh-k** [reply]
Thank you for looking into this! Super appreciated!

Unfortunately, no. I don't see any more details about the Headers Errors. I tried to Google what that error even means, but didn't find any useful information.

Here's a sample zip generated by my implementation of fflate:

[deploy-616b1c4bcb07117e1d0ed5fd.zip](https://github.com/101arrowz/fflate/files/7367266/deploy-616b1c4bcb07117e1d0ed5fd.zip)

You could generate a zip yourself if you were to create an account on Netlify and deploy a website there. Here's my repo: https://github.com/Hrishikesh-K/netlify-file-browser/tree/vite (Make sure to check the `vite` branch). The relevant code is in `/src/App.vue` - under `<script>` -> `methods` -> `downloadAll`.

**Reply by 101arrowz** [reply]
I've found the issue and it's actually quite interesting. Each file in a ZIP archive has two headers: one in the central directory at the end of the entire ZIP archive, and another local header right before the file in the archive. Most decompressors use the central directory header because it is at a defined position and you don't need to search for a magic number, but streaming decompressors can't read the end of the file so they use the local headers instead.

There's a bit called `STREAMED` in the bit flag for both headers that specifies whether or not the file was created in a streaming fashion (i.e. with `Zip` instead of `zipSync`). It is really only useful for streaming decompressors that need to find the end of each file in the ZIP without reading the central directory. Since streaming decompressors only use the local headers, I decided to only set the streaming bit for local headers but leave it out for the central directory to avoid some code complexity. Again, non-streaming decompressors read only the central directory but would not need to know whether the file is streamed or not, so setting the bit there is unnecessary, but streaming decompressors that read only the local headers do need to know, so the bit is set. In theory this decision should work fine.

It turns out 7-Zip in particular reads BOTH the local and central directory headers to cross-validate them, meaning it was confused by the mismatched bit flag and complained. However, the ZIP file was still perfectly valid, so it was able to decompress properly. Pretty much any other ZIP decompression tool will not cross-validate the headers and will instead only read one of them, meaning you won't get warnings from other tools.

Since this isn't really a bug and more a technicality on 7-Zip's part, I will fix this but will not publish a new version of `fflate` until there's another substantial change that needs to be made. So the fix will be available in v0.7.2, but that version may not come out until another bug is fixed.

**Reply by hrishikesh-k** [reply]
Wow, I'd never have imagined something as common as a zip would have so much to learn about. Thank you taking time to figure this out!

**Reply by dexgs** [reply]
> It turns out 7-Zip in particular reads BOTH the local and central directory headers to cross-validate them

Thanks for this insight. I'm doing something unrelated with Zip files, but I ran into this issue myself and I'm not sure I would have discovered the solution on my own.

**Reply by AlttiRi** [reply]
v0.7.3 still has Header Errors while unzippid with 7-zip.

**Reply by tamasfoldi** [reply]
v0.7.4 still has the Header Errors issue with 7-zip. @101arrowz was your fix applied in 0.7.2?

**Reply by 101arrowz** [reply]
It should be fixed now, are you sure you're using v0.7.4? If so please share your code so I can reproduce the issue.

**Reply by tamasfoldi** [reply]
You are right. It is fixed in 0.7.4. We have multiple modules and I've updated and tested two different modules 🤦
